zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResult.java

```diff
@@ -87,36 +89,51 @@ private String getData(String msg) {
     if (msg == null) {
       return null;
     }
-
-    Type[] types = Type.values();
-    for (Type t : types) {
-      String magic = "%" + t.name().toLowerCase();
-      if (msg.startsWith(magic + " ") || msg.startsWith(magic + "\n")) {
-        int magicLength = magic.length() + 1;
-        if (msg.length() > magicLength) {
-          return msg.substring(magicLength);
-        } else {
-          return "";
-        }
-      }
+    Type[] types = type.values();
+    TreeMap<Integer, Type> typesLastIndexInMsg = buildIndexMap(msg);
+    if (typesLastIndexInMsg.size() == 0) {
+      return msg;
+    } else {
+      Map.Entry<Integer, Type> lastType = typesLastIndexInMsg.lastEntry();
+      int magicLength = lastType.getValue().name().length() + 1; //add 1 for the %
+      int subStringPos = magicLength + lastType.getKey() + 1; // 1 for the last \w after magic
```

Comment:
Thanks for the question.
1) I fixed the comments to be more clear
2) a msg with only % will not get to that point as if
(typesLastIndexInMsg.size() == 0) which means we return  return msg;
3) added test cases for this in unit test

Eran

On Sat, Jul 25, 2015 at 10:16 PM Felix Cheung <notifications@github.com>
wrote:

> In
> zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResult.java
> <https://github.com/apache/incubator-zeppelin/pull/164#discussion_r35485972>
> :
>
> > -      if (msg.startsWith(magic + " ") || msg.startsWith(magic + "\n")) {
> > -        int magicLength = magic.length() + 1;
> > -        if (msg.length() > magicLength) {
> > -          return msg.substring(magicLength);
> > -        } else {
> > -          return "";
> > -        }
> > -      }
> > +    Type[] types = type.values();
> > +    TreeMap<Integer, Type> typesLastIndexInMsg = buildIndexMap(msg);
> > +    if (typesLastIndexInMsg.size() == 0) {
> > +      return msg;
> > +    } else {
> > +      Map.Entry<Integer, Type> lastType = typesLastIndexInMsg.lastEntry();
> > +      int magicLength = lastType.getValue().name().length() + 1; //add 1 for the %
> > +      int subStringPos = magicLength + lastType.getKey() + 1; // 1 for the last \w after magic
>
> \w? do you mean newline or regex:\w?
> if latter, what if someone types "%" - ie. no character following %
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-zeppelin/pull/164/files#r35485972>.
>